File: ESTUDOS/ESTRUTURAS-FLEXÍVEIS/LISTA-SIMPLES/Lista.java

public class Lista {
    
    // Class Celula
    public class Celula {
    
        // Atributos
        public int elemento;
        public Celula prox;

        // Construtores
        public Celula(){
            this(0);
        }

        public Celula(int x){
            this.elemento = x;
            this.prox = null;
        }

    }

    // Atributos da Lista Simples
    public Celula primeiro;
    public Celula ultimo;

    // Construtor da Lista Simples
    public Lista(){
        primeiro = new Celula();
        ultimo = primeiro;
    }

    public void InserirInicio(int x){
        Celula tmp = new Celula(x);
        tmp.prox = primeiro.prox;
        primeiro.prox = tmp;
        if (primeiro == ultimo){
            ultimo = tmp;
        }
        tmp = null;
    }

    public void InserirFim(int x){
        ultimo.prox = new Celula(x);
        ultimo = ultimo.prox;
    }
    }
    public void Inserir(int x, int pos){
        int tamanho = tamanho();
        if (pos < 0 || pos > tamanho){
            System.out.println("Erro!");
        } else if (pos == 0){
            InserirInicio(x);
        } else if (pos == tamanho){
            InserirFim(x);
        } else {
            Celula i = primeiro;
            for(int j = 0; j < pos; j++, i = i.prox);
            Celula tmp = new Celula(x);
            tmp.prox = i.prox;
            i.prox = tmp;
            i = tmp = null;
        }
    }
    public int RemoverInicio(){
    }
    public int RemoverFim(){
        if(primeiro == ultimo){
            System.out.println("Erro!");
        }
        Celula i;
        for(i = primeiro; i.prox != null; i = i.prox);
        int elemento = ultimo.elemento;
        ultimo = i;
        ultimo.prox = i = null;
        return elemento;
    }
    public int Remover(int pos) throws Exception{   
        int tamanho = tamanho();
        if (pos < 0 || pos > tamanho){
           throw new Exception("Erro!");
        } else if (pos == 0){
            RemoverInicio();
        } else if (pos == tamanho){
            RemoverFim();
        } else {
            Celula i = primeiro;
            for(int j = 0; j < pos; j++, i = i.prox);
            Celula tmp = i.prox;
            int elemento = i.elemento; 
            i.prox = tmp.prox;
            i = tmp = null;
            return elemento;
        }
    }


    public void Mostrar(){
        System.out.print("[");
        for(Celula i = primeiro; i != null; i = i.prox){
            System.out.print(i.elemento + " ");
        }
        System.out.print("]");
    }
    
    public int tamanho(){
        int tamanho = 0;
        for(Celula i = primeiro; i != ultimo; i = i.prox, tamanho++);
        return tamanho;
    }

}

File: ESTUDOS/ESTRUTURAS-FLEXÍVEIS/LISTA-SIMPLES/Lista.c
#include <stdio.h>
#include <stdlib.h>
#define bool   short
#define true   1
#define false  0

typedef struct Celula {
    int elemento;
    struct Celula *prox;
} Celula;

Celula* novaCelula (int x){
    Celula* nova = (Celula*)malloc(sizeof(Celula));
    nova->elemento = x;
    nova->prox = NULL;
    return nova;
}

Celula* primeiro;
Celula* ultimo; 

void start(){
    primeiro = novaCelula(-1);
    ultimo = primeiro;
}

int tamanho(){
    int tam = 0;
    Celula* i;
    for(i = primeiro; i->prox != NULL; i = i->prox, tam++);
    return tam;
}

void inserirInicio(int x){
    Celula* tmp = novaCelula(x);
    tmp->prox = primeiro->prox;
    primeiro->prox = tmp;
    if(primeiro == ultimo){
        ultimo = tmp;
    }
    tmp = NULL;
}

void inserirFim(int x){
    ultimo->prox = novaCelula(x);
    ultimo = ultimo->prox;
}

void inserir(int x, int pos){
    int tam = tamanho();
    if(pos < 0 || pos > tam){
        printf("Erro!");
    }
    else if (pos == 0){inserirInicio(x);}
    else if (pos == tamanho){inserirFim(x);}
    else{
        Celula* i = primeiro;
        for(int j = 0; j < pos; j++, i = i->prox);
        Celula* tmp = novaCelula(x);
        tmp->prox = i->prox;
        i->prox = tmp;
        tmp = i = NULL;
    }
}

int removerInicio(){
    if(primeiro == ultimo){
        printf("Erro");
    }
    Celula* tmp = primeiro;
    primeiro = primeiro->prox;
    int elemento = primeiro->elemento;
    tmp->prox = NULL;
    free(tmp);
    tmp = NULL;
    return elemento;
}

int removerFim(){
    if (primeiro == ultimo){
        printf("Erro");
    }
    Celula* i;
    for(i = primeiro; i->prox != ultimo; i = i->prox);
    int elemento = ultimo->elemento;
    ultimo = i;
    ultimo->prox = i = NULL;
    return elemento;
}

int remover(int pos){
    int elemento, tam = tamanho();
    if(pos < 0 || pos > tam){
        printf("Erro");
    }
    else if(pos == 0){removerInicio();}
    else if(pos == tam){removerFim();}
    else{
        Celula* i = primeiro;
        for(int j = 0; j < pos; j++, i = i->prox);
        Celula* tmp = i->prox;
        elemento = tmp->elemento;
        i->prox = tmp->prox;
        tmp->prox = NULL;
        free(tmp);
        tmp = NULL;
        return elemento;
    }
}

void mostrar(){
    printf("[");
    for(Celula* i = primeiro; i != NULL; i = i->prox){
        printf("%d", i->elemento);
    }
    printf("]");
}


